drivers/spi/spi_pico_pio.c

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (c) 2023 Stephen Boylan <stephen.boylan@beechwoods.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_spi_pio
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_pico_pio);
+
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "spi_context.h"
+
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+
+#include <hardware/pio.h>
+#include "hardware/clocks.h"
+
+#define PIO_CYCLES     (4)
+#define PIO_FIFO_DEPTH (4)
+
+struct spi_pico_pio_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *pin_cfg;
+	struct gpio_dt_spec clk_gpio;
+	struct gpio_dt_spec mosi_gpio;
+	struct gpio_dt_spec miso_gpio;
+};
+
+struct spi_pico_pio_data {
+	struct spi_context spi_ctx;
+	uint32_t tx_count;
+	uint32_t rx_count;
+	int cpol;
+	int cpha;
+	bool loop;
+	PIO pio;
+	size_t pio_sm;
+	int bits;
+	int dfs;
+	bool half_duplex;
+};
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_0_cpha_0, 0, 1,
+			    /*     .wrap_target */
+			    0x6101, /*  0: out    pins, 1         side 0 [1] */
+			    0x5101, /*  1: in     pins, 1         side 1 [1] */
+				/*     .wrap */
+);
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_1_cpha_1, 0, 2,
+			    /*     .wrap_target */
+			    0x7021, /*  0: out    x, 1            side 1 */
+			    0xa101, /*  1: mov    pins, x         side 0 [1] */
+			    0x5001, /*  2: in     pins, 1         side 1 */
+				/*     .wrap */
+);
+
+static float spi_pico_pio_clock_divisor(uint32_t spi_frequency)
+{
+	return (float)clock_get_hz(clk_sys) / (float)(PIO_CYCLES * spi_frequency);
+}
+
+static uint32_t spi_pico_pio_maximum_clock_frequency(void)
+{
+	return clock_get_hz(clk_sys) / PIO_CYCLES;
+}
+
+static uint32_t spi_pico_pio_minimum_clock_frequency(void)
+{
+	return clock_get_hz(clk_sys) / (PIO_CYCLES * 65536);
+}
```

Comment:
@soburi - clock-control is not supported on the Raspberry Pi Pico (or Pico W).  Those boards use a fixed clock.  However, looking back at this, it makes more sense to just use the clock frequency from the devicetree, so I've updated those functions to use that instead of calling into the Pico SDK and pushed the change to GitHub.  Thanks for pointing this out.